File: src/controller/zone.py
"""
Standalone zone-level read/write helpers for the MSI MPG B550 feature report packet.
"""

from typing import Dict, Any

from src.config import ZONE_OFFSETS, SYNC_SETTING_ONBOARD, SPEED_MEDIUM, BRIGHTNESS_100


def get_zone_info(packet: bytearray, zone_name: str) -> Dict[str, Any]:
    """Reads zone state from the 185-byte feature report packet."""
    offset = ZONE_OFFSETS[zone_name]
    effect = packet[offset]
    r1 = packet[offset + 1]
    g1 = packet[offset + 2]
    b1 = packet[offset + 3]
    flags = packet[offset + 4]
    speed = flags & 0x03
    brightness = (flags >> 2) & 0x1F
    r2 = packet[offset + 5]
    g2 = packet[offset + 6]
    b2 = packet[offset + 7]
    color_flags = packet[offset + 8]
    custom_color = bool(color_flags & 0x80)

    info = {
        "zone": zone_name,
        "effect": effect,
        "primary_rgb": (r1, g1, b1),
        "secondary_rgb": (r2, g2, b2),
        "speed": speed,
        "brightness": brightness,
        "custom_color": custom_color,
        "raw_flags": hex(flags),
        "raw_color_flags": hex(color_flags),
    }
    if zone_name in ("j_rainbow_1", "j_rainbow_2"):
        info["led_count_or_cycle"] = packet[offset + 10]
    return info


def set_zone_data(
    packet: bytearray,
    zone_name: str,
    mode: int,
    r: int,
    g: int,
    b: int,
    speed: int = SPEED_MEDIUM,
    brightness: int = BRIGHTNESS_100,
    led_count: int = 100,
) -> None:
    """Writes zone state into the 185-byte feature report packet."""
    offset = ZONE_OFFSETS[zone_name]
    packet[offset] = mode & 0xFF
    packet[offset + 1] = r & 0xFF
    packet[offset + 2] = g & 0xFF
    packet[offset + 3] = b & 0xFF
    packet[offset + 4] = ((brightness & 0x1F) << 2) | (speed & 0x03)
    packet[offset + 5] = r & 0xFF
    packet[offset + 6] = g & 0xFF
    packet[offset + 7] = b & 0xFF

    # Color flags: 0x80 indicates custom RGB color enabled
    # Onboard master zone also requires SYNC_SETTING_ONBOARD (0x01)
    if zone_name == "on_board_led":
        packet[offset + 8] = 0x80 | SYNC_SETTING_ONBOARD
    else:
        packet[offset + 8] = 0x80

    packet[offset + 9] = 0x00

    # Rainbow headers have the 11th byte: led_count / cycle
    if zone_name in ("j_rainbow_1", "j_rainbow_2"):
        packet[offset + 10] = led_count & 0xFF


__all__ = ["get_zone_info", "set_zone_data"]

File: src/controller/zone.rs
//! Standalone zone-level read/write helpers for the MSI MPG B550 feature report packet.
#![allow(dead_code)]

use crate::config::{
    get_zone_offset, SYNC_SETTING_ONBOARD,
};

#[derive(Debug, Clone)]
pub struct ZoneInfo {
    pub zone: String,
    pub effect: u8,
    pub primary_rgb: (u8, u8, u8),
    pub secondary_rgb: (u8, u8, u8),
    pub speed: u8,
    pub brightness: u8,
    pub custom_color: bool,
    pub raw_flags: u8,
    pub raw_color_flags: u8,
    pub led_count_or_cycle: Option<u8>,
}

pub fn get_zone_info(packet: &[u8; 185], zone_name: &str) -> Option<ZoneInfo> {
    let offset = get_zone_offset(zone_name)?;
    let effect = packet[offset];
    let r1 = packet[offset + 1];
    let g1 = packet[offset + 2];
    let b1 = packet[offset + 3];
    let flags = packet[offset + 4];
    let speed = flags & 0x03;
    let brightness = (flags >> 2) & 0x1F;
    let r2 = packet[offset + 5];
    let g2 = packet[offset + 6];
    let b2 = packet[offset + 7];
    let color_flags = packet[offset + 8];
    let custom_color = (color_flags & 0x80) != 0;

    let led_count_or_cycle = if zone_name == "j_rainbow_1" || zone_name == "j_rainbow_2" {
        Some(packet[offset + 10])
    } else {
        None
    };

    Some(ZoneInfo {
        zone: zone_name.to_string(),
        effect,
        primary_rgb: (r1, g1, b1),
        secondary_rgb: (r2, g2, b2),
        speed,
        brightness,
        custom_color,
        raw_flags: flags,
        raw_color_flags: color_flags,
        led_count_or_cycle,
    })
}

pub fn set_zone_data(
    packet: &mut [u8; 185],
    zone_name: &str,
    mode: u8,
    r: u8,
    g: u8,
    b: u8,
    speed: u8,
    brightness: u8,
    led_count: u8,
) -> bool {
    let offset = match get_zone_offset(zone_name) {
        Some(off) => off,
        None => return false,
    };

    packet[offset] = mode;
    packet[offset + 1] = r;
    packet[offset + 2] = g;
    packet[offset + 3] = b;
    packet[offset + 4] = ((brightness & 0x1F) << 2) | (speed & 0x03);
    packet[offset + 5] = r;
    packet[offset + 6] = g;
    packet[offset + 7] = b;

    if zone_name == "on_board_led" {
        packet[offset + 8] = 0x80 | SYNC_SETTING_ONBOARD;
    } else {
        packet[offset + 8] = 0x80;
    }

    packet[offset + 9] = 0x00;

    if zone_name == "j_rainbow_1" || zone_name == "j_rainbow_2" {
        packet[offset + 10] = led_count;
    }

    true
}
